_posts/2018-03-30-angular-security-best-practices.md

```diff
@@ -0,0 +1,151 @@
+    BROWSER_SANITIZATION_PROVIDERS
+    ...
+  ],
+  exports: [CommonModule, ApplicationModule]
+})
+export class BrowserModule {}
+{% endhighlight %}
+
+
+### The DOM sanitization service
+The goal of the DomSanitizer is to clean untrusted parts of values.
+
+The skeleton of the class looks like:
+{% highlight typescript %}
+export enum SecurityContext { NONE, HTML, STYLE, SCRIPT, URL, RESOURCE_URL }
+
+export abstract class DomSanitizer implements Sanitizer {
+  abstract sanitize(context: SecurityContext, value: SafeValue|string|null): string|null;
+  abstract bypassSecurityTrustHtml(value: string): SafeHtml;
+  abstract bypassSecurityTrustStyle(value: string): SafeStyle;
+  abstract bypassSecurityTrustScript(value: string): SafeScript;
+  abstract bypassSecurityTrustUrl(value: string): SafeUrl;
+  abstract bypassSecurityTrustResourceUrl(value: string): SafeResourceUrl;
+}
+{% endhighlight %}
+
+As you can see, there are two kinds of method patterns.
+The first one is the `sanitize` method, which gets the _context_ and an _untrusted value_ and returns a trusted value.
+The other ones are the `bypassSecurityTrustX` methods which are getting the _untrusted value_ according to the value usage and are returning a trusted object.
+
+#### The sanitize method
+If a value is trusted for the context, this sanitize method will (in case of a SafeValue) unwrap the contained safe value and use it directly.
+Otherwise, the value will be sanitized to be safe according to the security context.
+
+There are three main helper functions for sanitizing the values.
+The `sanitizeHtml` function sanitizes the untrusted HTML value by parsing the value and checks its tokens.
+The `sanitizeStyle` and `sanitizeUrl` functions sanitize the untrusted style or URL value by regular expressions.
+
+#### How can we disable the sanitization logic?
+In specific situations, it might be necessary to disable sanitization.
+Users can bypass security by constructing a value with one of the `bypassSecurityTrustX` methods, and then binding to that value from the template.
+
+> Be careful: If you trust a value that might be malicious, you are introducing a security vulnerability into your application!
+
+
+## 3. Content security policy (CSP)
+
+Content Security Policy is a defense-in-depth technique to prevent XSS.
+To enable CSP, configure your web server to return an appropriate `Content-Security-Policy` HTTP header.
+More info about specifying your CSP can be found on [Google Developers Page about Content Security Policy](https://developers.google.com/web/fundamentals/security/csp/){:target="blank" rel="noopener noreferrer"}.
+To check if your CSP is valid you can use the [CSP evaluator from google](https://csp-evaluator.withgoogle.com){:target="blank" rel="noopener noreferrer"}.
+
+
+## 4. Use the offline template compiler
+
+_Angular templates_ are the same as executable code: HTML, attributes, and binding expressions (but not the values bound) in templates are trusted to be safe.
```

Comment:
Offline compiler or AOT compiler? https://angular.io/guide/aot-compiler

Maybe add info how easy it is to enable --aot with angular cli